### DISCUSSION TITLE
How to view standalone header in certain preprocessor context

### DISCUSSION THREAD

**Discussion by recursivenomad** [clangd/clangd#2082] [Q&A] [answered]
So I'm interacting with a library which contains isolated header files (no `#include` directives in these headers, just some standalone `#define`s to be used elsewhere), but these header files contain `#ifdef` switches.  When viewing these files, all `#ifdef` switches are inactive.  This makes sense to me, because if this header is included at multiple stages of the program, there may not be a single state that these switches could be rendered in.

But I am curious, is it possible to view a generic header file in the context of a certain location it is called in the code, thus being able to visualize the active preprocessor switches when called from said location?

**Comment by HighCommander4** [comment, marked as answer]
> But I am curious, is it possible to view a generic header file in the context of a certain location it is called in the code, thus being able to visualize the active preprocessor switches when called from said location?

Not currently; this is basically https://github.com/clangd/clangd/issues/1534 (which in turn is a superset of https://github.com/clangd/clangd/issues/45).

That said, we can also think of this scenario as a bit of a special case of https://github.com/clangd/clangd/issues/45 in that the missing "context" to the header file is macro definitions (as opposed to general C/C++ declarations), and macro definitions have the property that they can be provided via the command line.

So, it's possible to employ a (hacky and manual) workaround by giving the header file an entry in `compile_commands.json` with the desired macro values:

```json
{
  "directory": "...",
  "file": "/path/to/header.hpp",
  "command": "clang -DMACRO1=VALUE1 -DMACRO2=VALUE2 /* other flags */"
}
```

which will cause clangd to open the header file in the context of those macro definitions.

**Reply by recursivenomad** [reply]
Thank you for the links and workaround, Nathan!
